src/compiler/emitter.ts

```diff
@@ -1639,10 +1639,12 @@ module ts {
             }
 
             if (root) {
+                currentSourceFile = root;
                 emit(root);

```

Comment:
It would also mean that the 'emit' method needs to check what type of node it has.  Something it doesn't do today.  It would also need to do this for the sourcemap/non-sourcemap case.  This is a simpler way to accomplish all that.